File: algorithms/tkf-2024-fall/p2/f.cpp
#include <iostream>
#include <map>
#include <deque>
using namespace std;

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(NULL);


    int n;
    cin >> n;

    deque<int> deq;
    map<int, int> pos;
    int out = 0;

    int cmd;
    int id;

    for (int i = 0; i < n; i++) {
        cin >> cmd;

        if (cmd == 1) {
            cin >> id;
            pos[id] = deq.size() + out;
            deq.push_back(id);
            continue;
        } 
        if (cmd == 2) {
            deq.pop_front();
            out = out + 1;
            continue;
        } 
        if (cmd == 3) {
            deq.pop_back();
            continue;
        }
        if (cmd == 4) {
            cin >> id;
            cout << pos[id] - out << endl;
            continue;
        }
        if (cmd == 5) {
            cout << deq.front() << endl;
        }
    }
}

File: algorithms/tkf-2024-fall/p2/f.py
from collections import deque

deq = deque()
pos = {}
out = 0

for _ in range(int(input())):
    cmd = input()
    if cmd[0] == '1':
        id = int(cmd[2:])
        pos[id] = len(deq) + out
        deq.append(id)
        continue
    if cmd[0] == '2':
        deq.popleft()
        out += 1
        continue
    if cmd[0] == '3':
        deq.pop()
        continue
    if cmd[0] == '4':
        id = int(cmd[2:])
        print(pos[id] - out)
    if cmd[0] == '5':
        print(deq[0])
        
    

    